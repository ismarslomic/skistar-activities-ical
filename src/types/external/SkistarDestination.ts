export enum SkistarDestination {
  'Hemsedal' = 'hemsedal',
  'Salen' = 'salen',
  'Ã…re' = 'are',
  'Vemdalen' = 'vemdalen',
  'Trysil' = 'trysil',
}
