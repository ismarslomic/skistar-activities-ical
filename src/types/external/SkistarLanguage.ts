export enum SkistarLanguage {
  'English' = 'en',
  'Norsk' = 'no',
  'Svenska' = 'sv',
  'Dansk' = 'da',
  'Deutsch' = 'de',
}
