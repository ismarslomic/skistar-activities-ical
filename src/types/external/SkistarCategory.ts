import { SkistarDestination } from './SkistarDestination'

export type SkistarCategory = {
  Id: string
  Destination?: SkistarDestination
  Name: string
  SortIndex: number
}
