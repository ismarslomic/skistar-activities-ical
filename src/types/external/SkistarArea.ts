import { SkistarCategory } from './SkistarCategory'
import { SkistarOccurrence } from './SkistarOccurrence'

export type SkistarArea = {
  Info: SkistarCategory
  Occurrences: SkistarOccurrence[]
}
