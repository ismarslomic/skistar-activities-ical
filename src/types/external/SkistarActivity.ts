import { SkistarCategory } from './SkistarCategory'
import { SkistarDestination } from './SkistarDestination'
import { SkistarArea } from './SkistarArea'

export type SkistarActivity = {
  Id: string
  Destination: SkistarDestination
  Heading: string
  Date: string
  DayOfMonth: string
  Month: string
  Times: string[]
  Description: string
  ImageUrl?: string
  Categories: SkistarCategory[]
  Tags: string[]
  MoreInfoUrl: null | string
  MoreInfoUrlText: null | string
  BookingUrl: null
  Areas: SkistarArea[]
  NumberOfOccurrences: number
  HasMultipleAdditionalTimes: boolean
  AdditionalTimesCount: null
}
