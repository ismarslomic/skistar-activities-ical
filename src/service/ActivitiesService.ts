import { SkistarFilterOptions } from '../types/external/SkistarFilterOptions'
import { SkistarArea } from '../types/external/SkistarArea'
import { SkistarCategory } from '../types/external/SkistarCategory'
import axios, { AxiosError, AxiosResponse } from 'axios'
import { SkistarActivity } from '../types/external/SkistarActivity'
import { SkistarOccurrence } from '../types/external/SkistarOccurrence'
import { ICalEventData } from 'ical-generator'
import { DateTime } from 'luxon'

export class ActivitiesService {
  static endpointUrl = 'https://www.skistar.com/__api/calendar/find'

  static async requestActivities(filterOptions: SkistarFilterOptions): Promise<ICalEventData[]> {
    console.debug(`Requesting activities from Skistar for filter ${JSON.stringify(filterOptions)}`)

    const requestBody = {
      fromDate: filterOptions.fromDate.toISODate(),
      toDate: filterOptions.toDate.toISODate(),
      destinationFilter: filterOptions.destinationFilter,
      areaFilter: filterOptions.areaFilter?.map((area: SkistarArea) => area.Info.Id),
      categoryFilter: filterOptions.categoryFilter?.map((category: SkistarCategory) => category.Id),
      skip: filterOptions.skip,
      take: filterOptions.take,
    }

    const config = {
      headers: {
        'Content-Type': 'application/json',
        'Accept-Language': filterOptions.language.valueOf(),
      },
    }

    return new Promise((resolve, reject) => {
      axios
        .post<SkistarActivity[]>(ActivitiesService.endpointUrl, requestBody, config)
        .then((response: AxiosResponse<SkistarActivity[]>) => {
          console.debug('Reading successfully activities from Skistar API.')

          const events: ICalEventData[] = []
          response.data.forEach((skistarActivity: SkistarActivity) =>
            events.push(...mapToCalendarEvent(skistarActivity))
          )

          resolve(events)
        })
        .catch((error: AxiosError) => {
          console.error('Reading activities from Skistar API failed', error)
          reject(error.message)
        })
    })

    function mapToCalendarEvent(activity: SkistarActivity): ICalEventData[] {
      const localDateString: string = activity.Date // Måndag 17 april 2023
      const summary: string = activity.Heading
      const description: string = activity.Description
      const events: ICalEventData[] = []
      const imageUrl: string | null = activity.ImageUrl

      activity.Areas.forEach((activityArea: SkistarArea) => {
        const destinationName: string = activityArea.Info.Name
        const uniqueOccurrences: SkistarOccurrence[] = getUniqueOccurrences(activityArea.Occurrences)

        uniqueOccurrences.forEach((occurrence: SkistarOccurrence) => {
          let startTime: DateTime
          let endTime: DateTime

          if (occurrence.StartTime != null && occurrence.EndTime != null) {
            startTime = ActivitiesService.cestDateTime(localDateString, occurrence.StartTime)
            endTime = ActivitiesService.cestDateTime(localDateString, occurrence.EndTime)
          } else {
            // handles whole day events
            startTime = ActivitiesService.cestDateTime(localDateString, '00:00')
            endTime = ActivitiesService.cestDateTime(localDateString, '00:00')
          }

          const isWholeDay: boolean = occurrence.IsWholeDay
          const location: string = occurrence.LocationDescription ? occurrence.LocationDescription : destinationName

          const event: ICalEventData = {
            start: startTime,
            end: endTime,
            summary: summary,
            description: description,
            location: location,
            allDay: isWholeDay,
            attachments: imageUrl ? [imageUrl] : [],
          }

          events.push(event)
        })
      })

      return events
    }

    // Remove duplicate occurrences with the same start and end time
    function getUniqueOccurrences(occurrences: SkistarOccurrence[]): SkistarOccurrence[] {
      return occurrences.filter((occurrence: SkistarOccurrence, index: number): boolean => {
        return (
          occurrences.findIndex(
            (otherOccurrence: SkistarOccurrence) =>
              otherOccurrence.StartTime === occurrence.StartTime &&
              otherOccurrence.EndTime === occurrence.EndTime &&
              otherOccurrence.LocationDescription === occurrence.LocationDescription
          ) === index
        )
      })
    }
  }

  /**
   *
   * @param date as a string in the format "cccc d LLLL yyyy" in swedish, example "Lördag 1 april 2023"
   * @param time as a string in the format "HH:mm", example "10:00"
   * @returns a DateTime object in CEST timezone, example "2023-04-01T10:00:00+02:00"
   */
  static cestDateTime(date: string, time: string): DateTime {
    const localDateTime = `${date} ${time} Europe/Oslo`

    // Example: Måndag 17 april 2023 10:00 Europe/Oslo
    const localFormat = 'cccc d LLLL yyyy HH:mm z'

    return DateTime.fromFormat(localDateTime, localFormat, { locale: 'sv' })
  }
}
