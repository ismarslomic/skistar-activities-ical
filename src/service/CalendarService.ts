import ical, { ICalCalendar, ICalEventData } from 'ical-generator'
import { ActivitiesService } from './ActivitiesService'
import { DateTime } from 'luxon'
import { getVtimezoneComponent } from '@touch4it/ical-timezones'
import * as http from 'http'
import { SkistarDestination } from '../types/external/SkistarDestination'
import { SkistarLanguage } from '../types/external/SkistarLanguage'
import { logger } from '../logging/logger'

export class CalendarService {
  private readonly daysInFuture: number
  private readonly destination: SkistarDestination
  private readonly language: SkistarLanguage
  private readonly calFileName: string

  constructor(daysInFuture: number, destination: SkistarDestination, language: SkistarLanguage, calFileName: string) {
    this.daysInFuture = daysInFuture
    this.destination = destination
    this.language = language
    this.calFileName = calFileName
  }

  async createCalendar(): Promise<ICalCalendar> {
    try {
      const activityEvents: ICalEventData[] = await ActivitiesService.requestActivities({
        fromDate: DateTime.now(),
        toDate: DateTime.now().plus({ days: this.daysInFuture }),
        destinationFilter: this.destination,
        language: this.language,
      })
      logger.info(`Retrieved ${activityEvents.length} activity events from Skistar`)

      const destinationName = this.getDestinationName(this.destination)

      return ical({
        name: `Skistar activities in ${destinationName}`,
        events: activityEvents,
        prodId: { company: 'skistar.com', product: 'activities', language: 'NO' },
        timezone: {
          name: ActivitiesService.timeZone,
          generator: getVtimezoneComponent,
        },
      })
    } catch (e) {
      logger.error('Failed to retrieve activities from Skistar API and create the calendar', e)

      return ical({
        name: 'Skistar activities in Hemsedal',
        events: [],
        prodId: { company: 'skistar.com', product: 'activities', language: 'NO' },
        timezone: {
          name: 'Europe/Oslo',
          generator: getVtimezoneComponent,
        },
      })
    }
  }

  getDestinationName(value: string) {
    const indexOfValue = Object.values(SkistarDestination).indexOf(value as unknown as SkistarDestination)
    return Object.keys(SkistarDestination)[indexOfValue]
  }

  async serveCalendar(res: http.ServerResponse): Promise<ICalCalendar> {
    const calendar: ICalCalendar = await this.createCalendar()
    return calendar.serve(res, this.calFileName)
  }
}
