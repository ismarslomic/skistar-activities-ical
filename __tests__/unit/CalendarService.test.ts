import nock = require('nock')
import { ActivitiesService } from '../../src/service/ActivitiesService'
import * as mockResponse from '../../__mocks__/activitiesResponseMock'
import { CalendarService } from '../../src/service/CalendarService'
import { SkistarDestination } from '../../src/types/external/SkistarDestination'
import { SkistarLanguage } from '../../src/types/external/SkistarLanguage'
import { ICalCalendar } from 'ical-generator'
import { v4 } from 'uuid'
import { Settings } from 'luxon'

jest.mock('uuid', () => ({ v4: jest.fn() }))

describe('CalendarService createCalendar', () => {
  afterEach(() => {
    nock.cleanAll()
    jest.clearAllMocks()
    Settings.now = () => new Date().valueOf()
  })

  describe('given 200 response with non empty list with activities', () => {
    beforeEach(() => {
      // mock HTTP request
      nock(ActivitiesService.endpointUrl).post('').reply(200, mockResponse.activities)
      // mock v4.uuid()
      ;(v4 as jest.Mock).mockReturnValue('mock-uuid-1234')
      // mock Luxon DateTime.now()
      Settings.now = () => new Date('2023-04-24T13:00:00.000+02:00').valueOf()
    })

    test('calendar with 18 events are returned', async () => {
      const calendarService: CalendarService = new CalendarService(
        14,
        SkistarDestination.Ã…re,
        SkistarLanguage.Norsk,
        'calendar.ics'
      )

      const calendar: ICalCalendar = await calendarService.createCalendar()
      expect(calendar.events().length).toBe(18)

      expect(calendar.toString()).toMatchSnapshot()
    })
  })

  describe('given 404 not found response with error message', () => {
    beforeEach(() => {
      nock(ActivitiesService.endpointUrl).post('').reply(404)
    })

    test('calendar with empty list of events is returned', async () => {
      const calendarService: CalendarService = new CalendarService(
        14,
        SkistarDestination.Hemsedal,
        SkistarLanguage.Norsk,
        'calendar.ics'
      )

      const calendar: ICalCalendar = await calendarService.createCalendar()
      expect(calendar.events().length).toBe(0)
    })
  })
})
