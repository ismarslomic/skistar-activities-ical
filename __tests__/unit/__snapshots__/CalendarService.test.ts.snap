// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarService createCalendar given 200 response with non empty list with activities calendar with 18 events are returned 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//skistar.com//activities//NO
NAME:Skistar activities in Åre
X-WR-CALNAME:Skistar activities in Åre
BEGIN:VTIMEZONE
TZID:Europe/Oslo
TZURL:http://tzurl.org/zoneinfo-outlook/Europe/Oslo
X-LIC-LOCATION:Europe/Oslo
BEGIN:DAYLIGHT
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
TZNAME:CEST
DTSTART:19700329T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
TZNAME:CET
DTSTART:19701025T030000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
END:VTIMEZONE
TIMEZONE-ID:Europe/Oslo
X-WR-TIMEZONE:Europe/Oslo
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230313T133000
DTEND:20230313T140000
SUMMARY:Meet Valle
LOCATION:Valle's stage\\, Children's area
DESCRIPTION:Say hello to Valle and get to know him even better. It is a pe
 rfect opportunity to ask questions\\, take a picture or hear about his adve
 ntures.
ATTACH:https://www.skistar.com/calendar-api/assets/18/18d3a111-fd30-40b2-b
 277-2e625bfffef8/img8981ret.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART;VALUE=DATE:20230313
DTEND;VALUE=DATE:20230313
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
SUMMARY:Climbing and rappelling at Fjellkafeen
LOCATION:At Fjellkafeen
DESCRIPTION:Suitable for everyone from about 7 years and up
ATTACH:https://www.skistar.com/calendar-api/assets/7e/7e4414b1-0c15-4fe9-b
 472-dc29ebeeea9d/hemsedalklattring.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART;VALUE=DATE:20230313
DTEND;VALUE=DATE:20230313
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
SUMMARY:Climbing and rappelling at Fjellkafeen
LOCATION:At Fjellkafeen
DESCRIPTION:Suitable for everyone from about 7 years and up
ATTACH:https://www.skistar.com/calendar-api/assets/7e/7e4414b1-0c15-4fe9-b
 472-dc29ebeeea9d/hemsedalklattring.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230314T090000
DTEND:20230314T190000
SUMMARY:Non-stop skiing until 19:00
LOCATION:Hemsedal skisenter
DESCRIPTION:Lifts are open from 09:00 until 19:00.
ATTACH:https://www.skistar.com/calendar-api/assets/43/43a85cd9-1f4e-4eb7-a
 462-653f5c1b0d7a/190408omhemsedal-31169lite.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230314T133000
DTEND:20230314T140000
SUMMARY:Valle's Adventure
LOCATION:Valle's stage\\, Children's area
DESCRIPTION:When Valle is on the go\\, you never know what may happen. Join
  an activity with Valle\\, such as dancing stop and games.
ATTACH:https://www.skistar.com/calendar-api/assets/4e/4ec22916-6a45-4f5b-b
 ba1-453ac9d0ef03/img8950.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230314T151500
DTEND:20230314T154500
SUMMARY:Valle's Show
LOCATION:Valle's stage by the children's area
DESCRIPTION:Valle and his friends offer family entertainment in the form o
 f singing and dancing\\, where learning is the common theme.
ATTACH:https://www.skistar.com/calendar-api/assets/d8/d8846086-7d8e-492c-a
 e2c-fb0c94bd3d32/412326220200680f87275k.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230315T090000
DTEND:20230315T190000
SUMMARY:Non-stop skiing until 19:00
LOCATION:Hemsedal skisenter
DESCRIPTION:Lifts are open from 09:00 until 19:00.
ATTACH:https://www.skistar.com/calendar-api/assets/43/43a85cd9-1f4e-4eb7-a
 462-653f5c1b0d7a/190408omhemsedal-31169lite.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230315T141500
DTEND:20230315T144500
SUMMARY:Go skiing with Valle
LOCATION:Lift A1 Valle 1\\, Children's Area
DESCRIPTION:Valle is skiing with the children in the children's area. Give
  Valle a hug\\, wave and say hello!
ATTACH:https://www.skistar.com/calendar-api/assets/3a/3a84bf66-8e14-43ca-8
 1f3-46bd86b99f02/skiing-valle.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230315T163000
DTEND:20230315T180000
SUMMARY:Sledding in Fjellheisen
LOCATION:Hemsedal skisenter
DESCRIPTION:With a headlamp and snowracer\\, we sled down the mountain.
ATTACH:https://www.skistar.com/calendar-api/assets/b6/b6a93634-8333-4e00-b
 7b0-7d877672e5bc/aking.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230315T183000
DTEND:20230315T200000
SUMMARY:Sledding from top to bottom
LOCATION:Hemsedal skisenter
DESCRIPTION:Sledding from top to bottom
ATTACH:https://www.skistar.com/calendar-api/assets/1b/1b2016a8-55da-400c-b
 5cf-2d9ef91d1965/topp-til-bunn.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230316T110000
DTEND:20230316T113000
SUMMARY:Meet Valle
LOCATION:Valle's stage\\, Children's area
DESCRIPTION:Say hello to Valle and get to know him even better. It is a pe
 rfect opportunity to ask questions\\, take a picture or hear about his adve
 ntures.
ATTACH:https://www.skistar.com/calendar-api/assets/18/18d3a111-fd30-40b2-b
 277-2e625bfffef8/img8981ret.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230316T140000
DTEND:20230316T143000
SUMMARY:Valle's Show
LOCATION:Valle's stage by the children's area
DESCRIPTION:Valle and his friends offer family entertainment in the form o
 f singing and dancing\\, where learning is the common theme.
ATTACH:https://www.skistar.com/calendar-api/assets/d8/d8846086-7d8e-492c-a
 e2c-fb0c94bd3d32/412326220200680f87275k.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230316T160000
DTEND:20230316T210000
SUMMARY:Indoor climbing
LOCATION:SkiStar Lodge Hemsedal Suites
DESCRIPTION:Hemsedal climbing center is located in SkiStar Lodge Hemsedal 
 Suites in Hemsedal ski center. The center consists of a 14 meter high clim
 bing wall and a large boulder area.
ATTACH:https://www.skistar.com/calendar-api/assets/8a/8ab11984-a76a-4693-b
 5c4-50200ce8f5a6/klatring-inne.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230317T110000
DTEND:20230317T113000
SUMMARY:Go skiing with Valle
LOCATION:Lift A1 Valle 1\\, Children's Area
DESCRIPTION:Valle is skiing with the children in the children's area. Give
  Valle a hug\\, wave and say hello!
ATTACH:https://www.skistar.com/calendar-api/assets/3a/3a84bf66-8e14-43ca-8
 1f3-46bd86b99f02/skiing-valle.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230317T180000
DTEND:20230317T193000
SUMMARY:Sledding in Fjellheisen
LOCATION:Hemsedal skisenter
DESCRIPTION:With a headlamp and snowracer\\, we sled down the mountain.
ATTACH:https://www.skistar.com/calendar-api/assets/b6/b6a93634-8333-4e00-b
 7b0-7d877672e5bc/aking.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230317T180000
DTEND:20230317T210000
SUMMARY:Indoor climbing
LOCATION:SkiStar Lodge Hemsedal Suites
DESCRIPTION:Hemsedal climbing center is located in SkiStar Lodge Hemsedal 
 Suites in Hemsedal ski center. The center consists of a 14 meter high clim
 bing wall and a large boulder area.
ATTACH:https://www.skistar.com/calendar-api/assets/8a/8ab11984-a76a-4693-b
 5c4-50200ce8f5a6/klatring-inne.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230317T180000
DTEND:20230317T203000
SUMMARY:Evening skiing
LOCATION:Hemsedal skisenter
DESCRIPTION:Evening skiing at 1800-2030
ATTACH:https://www.skistar.com/calendar-api/assets/69/695b1ed3-e293-4b89-8
 d12-9a0a55135f4b/kveldsskihemsedal.jpg
END:VEVENT
BEGIN:VEVENT
UID:mock-uuid-1234
SEQUENCE:0
DTSTAMP:20230424T130000
DTSTART:20230317T200000
DTEND:20230317T213000
SUMMARY:Sledding from top to bottom
LOCATION:Hemsedal skisenter
DESCRIPTION:Sledding from top to bottom
ATTACH:https://www.skistar.com/calendar-api/assets/1b/1b2016a8-55da-400c-b
 5cf-2d9ef91d1965/topp-til-bunn.jpg
END:VEVENT
END:VCALENDAR"
`;
